src/main/java/slimeknights/tconstruct/tables/tileentity/table/tinkerstation/TinkerStationTileEntity.java

```diff
@@ -0,0 +1,215 @@
+import slimeknights.tconstruct.common.Sounds;
+import slimeknights.tconstruct.library.network.TinkerNetwork;
+import slimeknights.tconstruct.library.recipe.RecipeTypes;
+import slimeknights.tconstruct.library.recipe.tinkerstation.ITinkerStationRecipe;
+import slimeknights.tconstruct.shared.inventory.ConfigurableInvWrapperCapability;
+import slimeknights.tconstruct.tables.TinkerTables;
+import slimeknights.tconstruct.tables.inventory.table.tinkerstation.TinkerStationContainer;
+import slimeknights.tconstruct.tables.network.UpdateCraftingRecipePacket;
+import slimeknights.tconstruct.tables.network.UpdateTinkerStationRecipePacket;
+import slimeknights.tconstruct.tables.tileentity.crafting.CraftingInventoryWrapper;
+import slimeknights.tconstruct.tables.tileentity.crafting.LazyResultInventory;
+import slimeknights.tconstruct.tables.tileentity.table.RetexturedTableTileEntity;
+
+import javax.annotation.Nullable;
+import java.util.Collections;
+
+public class TinkerStationTileEntity extends RetexturedTableTileEntity implements LazyResultInventory.ILazyCrafter {
+
+  public static final int TINKER_SLOT = 5;
+  public static final int OUTPUT_SLOT = 0;
+
+  /** Last crafted crafting recipe */
+  @Nullable
+  private ITinkerStationRecipe lastRecipe;
+  /** Result inventory, lazy loads results */
+  @Getter
+  private final LazyResultInventory craftingResult;
+  /** Crafting inventory for the recipe calls */
+  private final TinkerStationInventoryWrapper inventoryWrapper;
+
+  public TinkerStationTileEntity() {
+    super(TinkerTables.tinkerStationTile.get(), "gui.tconstruct.tinker_station", 6);
+    this.itemHandler = new ConfigurableInvWrapperCapability(this, false, false);
+    this.itemHandlerCap = LazyOptional.of(() -> this.itemHandler);
+    this.inventoryWrapper = new TinkerStationInventoryWrapper(this);
+    this.craftingResult = new LazyResultInventory(this);
+  }
+
+  @Nullable
+  @Override
+  public Container createMenu(int menuId, PlayerInventory playerInventory, PlayerEntity playerEntity) {
+    return new TinkerStationContainer(menuId, playerInventory, this);
+  }
+
+  /* Crafting */
+
+  @Override
+  public ItemStack calcResult() {
+    if (this.world == null) {
+      return ItemStack.EMPTY;
+    }
+
+    // assume empty unless we learn otherwise
+    ItemStack result = ItemStack.EMPTY;
+
+    if (!this.world.isRemote && this.world.getServer() != null) {
+      RecipeManager manager = this.world.getServer().getRecipeManager();
+
+      // first, try the cached recipe
+      ITinkerStationRecipe recipe = lastRecipe;
+      // if it does not match, find a new recipe
+      if (recipe == null || !recipe.matches(this.inventoryWrapper, this.world)) {
+        recipe = manager.getRecipe(RecipeTypes.TINKER_STATION, this.inventoryWrapper, this.world).orElse(null);
+      }
+
+      // if we have a recipe, fetch its result
+      if (recipe != null) {
+        result = recipe.getCraftingResult(this.inventoryWrapper);
```

Comment:
We need to call validate at some point to prevent the result if the recipe does not validate. Maybe here? We really need the result of validate in the screen, could always send it in a packet from this method, cache it for new clients to open the container